### DISCUSSION TITLE
Open Cef using Form.ShowDialog and avoid hanging or app  unresponsiveness

### DISCUSSION THREAD

**Discussion by SorceryAdmin** [cefsharp/CefSharp#3418] [Show and tell]
**Fix or  WorkAround Report not sure how to call it.**

**What version of the product are you using?**
Tested from 79 to 88-pre using Nuget

**What architecture x86 or x64?**
x86 and x64

**On what operating system?**
Win10 and Window8

**Are you using WinForms, WPF or OffScreen?**
WinForms

**What Framework?**
.net 4.5,4.62,4.7, Net Core 3/3.1 and Net 5

**What is the expected output? What do you see instead?**

When you use a a Window Form with CEF and then  you can open a new form( as a pop up  ) the main form get stuck when you close the child window. Tried with regular show or custom ILifeSpanHandler and still does not work.
similar to what his post experience:
https://github.com/cefsharp/CefSharp/issues/1574
https://github.com/cefsharp/CefSharp/issues/2086
https://groups.google.com/g/cefsharp/c/3Mi0PY6r22s

How to reproduce the error we are fixing:
on a windows form that has a CefSharpBrowser add a button and add code  so when that button is hit we  open a new winform using this :

```
var cefForPdfView = new CefSharp.WinForms.ChromiumWebBrowser("about:blank", new CefSharp.RequestContext());
Form previewForm = new Form();
previewForm.Text = fileUrl;          
previewForm.Size = new Size(1024,768);
cefForPdfView.Dock = DockStyle.Fill;
previewForm.Controls.Add(cefForPdfView);
cefForPdfView.IsBrowserInitializedChanged += ((object sender, EventArgs e) =>
{
  var cefForPdfView = ((CefSharp.WinForms.ChromiumWebBrowser)sender);
  if (cefForPdfView.IsBrowserInitialized)
	cefForPdfView.Load("http://ipchicken.com" /*Generated Proof Of Registered URL:fileUrl*/);      			
});
previewForm.ShowDialog();
```
When that `previewForm` is Closed/Disposed the Main Form in fact the whole App (if other pop up windows are open) whill hang and be unresponsive for about 30 seconds more less will depend.

To avoid that error the updated code looks like this:
```
var cefForPdfView = new CefSharp.WinForms.ChromiumWebBrowser("about:blank", new CefSharp.RequestContext());
Form previewForm = new Form();
previewForm.Text = fileUrl;                
previewForm.Size = new Size(1024,768);
previewForm.TopLevel = true;
previewForm.CreateControl();     
cefForPdfView.Dock = DockStyle.Fill;
previewForm.Controls.Add(cefForPdfView);
cefForPdfView.IsBrowserInitializedChanged += ((object sender, EventArgs e) =>
{
	var cefForPdfView = ((CefSharp.WinForms.ChromiumWebBrowser)sender);
	if (cefForPdfView.IsBrowserInitialized)
	{
		cefForPdfView.Load("http://ipchicken.com"/*Generated Proof Of Registered URL:fileUrl*/);              
	}
});
this.AddOwnedForm(previewForm);
if (this.InvokeRequired)
	this.Invoke(new Action(() => { previewForm.Show(); }));
else
	previewForm.Show();
```
Hope this help others cause was gaving me a pain for a system where multiple pop ups by user where needed. 
Tried threads and a lot of complicated ideas and was this simple one the charm.

**Comment by amaitland** [comment]
ShowDialog is a special case, when showing a modal dialog a modal message loop is created. Chromium isn't designed to be run in such a message loop.

For reference 
https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Form.cs,6224
https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Application.cs,3338
https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Application.cs,2718

I don't remember the specifics, just that there are problems.
